doc/userguide/rules/rules-internals.rst

```diff
@@ -0,0 +1,222 @@
+Detect Engine will match against - if applicable:
+
+#. IPOnly rules;
+#. Packet/payload related rules;
+#. Frame keywords;
+#. Application layer protocol transactions.
+
+During packet inspection, if the signature uses the last two, inspection is
+left to those steps.
+
+Each rule that matches becomes a ``PacketAlert``. These will, after all matches
+for the packet have been processed, become the alerts that are seen in Suricata
+logs.
+
+Considerations on inspection steps
+----------------------------------
+
+IPOnly rules
+~~~~~~~~~~~~
+
+Without optimization, IPOnly signatures would match on every packet on a flow.
+To improve performance, what Suricata does is to evaluate rules that are ``ip-
+only`` only once per flow direction, for the first packet in each direction.
+
+Application layer protocol transactions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Each parser has its own state machine, where it uses a per direction parsing
+state "progress". For each progress value keywords can be registered. So, for
+instance, ``http`` has a value "request line available" for which there are
+keywords like :ref:`http.uri <rules-http-uri-normalization>`, :ref:`http.method
+<http.method>` etc. registered. If in the traffic the Engine reaches this state,
+the signatures with those keywords may be already evaluated, even if they have a
+lower priority than a body inspecting signature.
+
+Implications
+------------
+
+Use-case: action precedence and interaction with ``ip-only`` rules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To illustrate what may be counter-intuitive implications of how inspection
+steps, action prioritization and rule keywords interact and affect the engine
+behavior, we will use a real case example. Consider these three rules:
+
+.. container:: example-rule
+
+    pass tcp 0.0.0.0/0 any <> 0.0.0.0/0 443 (msg:"Allow TCP in port 443"; flow:
+    not_established; sid:1; rev:1;)
+
+.. container:: example-rule
+
+    pass tcp 0.0.0.0/0 any <> 0.0.0.0/0 80 (msg:"Allow TCP in port 80"; flow:
+    not_established; sid:2; rev:1;)
+
+.. container:: example-rule
+
+    drop ip 0.0.0.0/0 any -> 0.0.0.0/0 any (msg:"No outbound internet access
+    from host"; sid:3; rev:1;)
+
+The first two are signatures that analyze individual packets and match only if
+the flow has not been established (``flow:not_established``): the rules grant
+``PASS`` to the matched packet - but not to its flow.
+
+The third signature is considered ``ip-only``. This means it will be evaluated
+for the *first* packet in both directions of a flow, in addition to rules 1 and
+2. By extension, **the other packets in the same flow will not be evaluated
+against this rule.**
+
+With an action order configuration that prioritizes ``PASS`` over ``DROP``, this
+means that rules 1 and 2 will have a higher internal priority over rule 3,
+therefore nullifying the ``DROP`` outcome. The result: a flow expected to be
+dropped might not be.
```

Comment:
I think this bit needs more explanation, especially with regards to what sort of flow may not be dropped in the scenario described...